sdk/identity/Azure.Identity/src/ImdsManagedIdentitySource.cs

```diff
@@ -0,0 +1,132 @@
+            }
+
+            if (available)
+            {
+                AzureIdentityEventSource.Singleton.ImdsEndpointFound(s_imdsEndpoint);
+            }
+            else
+            {
+                AzureIdentityEventSource.Singleton.ImdsEndpointUnavailable(s_imdsEndpoint);
+            }
+
+            return available ? new ImdsManagedIdentitySource(pipeline, clientId) : default;
+        }
+
+        internal ImdsManagedIdentitySource(HttpPipeline pipeline, string clientId)
+        {
+            _pipeline = pipeline;
+            _clientId = clientId;
+        }
+
+        public Request CreateRequest(string[] scopes)
+        {
+            if (_identityUnavailableErrorMessage != default)
+            {
+                throw new CredentialUnavailableException(_identityUnavailableErrorMessage);
+            }
+
+            // covert the scopes to a resource string
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+
+            Request request = _pipeline.CreateRequest();
+            request.Method = RequestMethod.Get;
+            request.Headers.Add("Metadata", "true");
+            request.Uri.Reset(s_imdsEndpoint);
+            request.Uri.AppendQuery("api-version", ImdsApiVersion);
+
+            request.Uri.AppendQuery("resource", resource);
+
+            if (!string.IsNullOrEmpty(_clientId))
+            {
+                request.Uri.AppendQuery("client_id", _clientId);
+            }
+
+            return request;
+        }
+
+        public AccessToken GetAccessTokenFromJson(in JsonElement jsonAccessToken, in JsonElement jsonExpiresOn)
+        {
+            // the seconds from epoch may be returned as a Json number or a Json string which is a number
+            // depending on the environment.  If neither of these are the case we throw an AuthException.
+            if (jsonExpiresOn.ValueKind == JsonValueKind.Number && jsonExpiresOn.TryGetInt64(out long expiresOnSec) ||
+                jsonExpiresOn.ValueKind == JsonValueKind.String && long.TryParse(jsonExpiresOn.GetString(), out expiresOnSec))
+            {
+                return new AccessToken(jsonAccessToken.GetString(), DateTimeOffset.FromUnixTimeSeconds(expiresOnSec));
+            }
+
+            throw new AuthenticationFailedException(ManagedIdentityClient.AuthenticationResponseInvalidFormatError);
+        }
+
+        public async ValueTask HandleFailedRequestAsync(Response response, ClientDiagnostics diagnostics, bool async)
+        {
+            if (response.Status == 400)
+            {
+                string message = _identityUnavailableErrorMessage ?? await diagnostics
+                    .CreateRequestFailedMessageAsync(response, IdentityUnavailableError, null, null, async)
+                    .ConfigureAwait(false);
+
+                Interlocked.CompareExchange(ref _identityUnavailableErrorMessage, message, null);
+                throw new CredentialUnavailableException(message);
```

Comment:
Looks like the credential will send another request the next time a token is requested. Is this a change from the prior behavior?

Reply:
No, error message is stored in `_identityUnavailableErrorMessage` and will be used later to throw exception: https://github.com/Azure/azure-sdk-for-net/pull/14770/files/807bb81803ec86e460538d8c76d101ff57f68c9f#diff-7bbc16093e3640db68eb0921760088e1R84